services/archiver/archiver.go

```diff
@@ -0,0 +1,309 @@
+	archivePath     string
+	archiveType     git.ArchiveType
+	archiveComplete bool
+	commit          *git.Commit
+}
+
+var archiveInProgress []*ArchiveRequest
+var archiveMutex sync.Mutex
+var archiveCond *sync.Cond
+
+// These facilitate testing, by allowing the unit tests to control (to some extent)
+// the goroutine used for processing the queue.
+var archiveQueueMutex *sync.Mutex
+var archiveQueueStartCond *sync.Cond
+var archiveQueueReleaseCond *sync.Cond
+
+// GetArchivePath returns the path from which we can serve this archive.
+func (aReq *ArchiveRequest) GetArchivePath() string {
+	return aReq.archivePath
+}
+
+// GetArchiveName returns the name of the caller, based on the ref used by the
+// caller to create this request.
+func (aReq *ArchiveRequest) GetArchiveName() string {
+	return aReq.refName + aReq.ext
+}
+
+// IsComplete returns the completion status of this request.
+func (aReq *ArchiveRequest) IsComplete() bool {
+	return aReq.archiveComplete
+}
+
+// The caller must hold the archiveMutex across calls to getArchiveRequest.
+func getArchiveRequest(repo *git.Repository, commit *git.Commit, archiveType git.ArchiveType) *ArchiveRequest {
+	for _, r := range archiveInProgress {
+		// Need to be referring to the same repository.
+		if r.repo.Path == repo.Path && r.commit.ID == commit.ID && r.archiveType == archiveType {
+			return r
+		}
+	}
+	return nil
+}
+
+// DeriveRequestFrom creates an archival request, based on the URI.  The
+// resulting ArchiveRequest is suitable for being passed to ArchiveRepository()
+// if it's determined that the request still needs to be satisfied.
+func DeriveRequestFrom(ctx *context.Context, uri string) *ArchiveRequest {
+	if ctx.Repo == nil || ctx.Repo.GitRepo == nil {
+		log.Trace("Repo not initialized")
+		ctx.Error(404)
+		return nil
+	}
+	r := &ArchiveRequest{
+		uri:  uri,
+		repo: ctx.Repo.GitRepo,
+	}
+
+	switch {
+	case strings.HasSuffix(uri, ".zip"):
+		r.ext = ".zip"
+		r.archivePath = path.Join(r.repo.Path, "archives/zip")
+		r.archiveType = git.ZIP
+	case strings.HasSuffix(uri, ".tar.gz"):
+		r.ext = ".tar.gz"
+		r.archivePath = path.Join(r.repo.Path, "archives/targz")
+		r.archiveType = git.TARGZ
+	default:
+		log.Trace("Unknown format: %s", uri)
+		ctx.Error(404)
+		return nil
+	}
+
+	r.refName = strings.TrimSuffix(r.uri, r.ext)
+	if !com.IsDir(r.archivePath) {
+		if err := os.MkdirAll(r.archivePath, os.ModePerm); err != nil {
+			ctx.ServerError("Download -> os.MkdirAll(archivePath)", err)
+			return nil
+		}
+	}
+
+	// Get corresponding commit.
+	var (
+		err error
+	)
+	if r.repo.IsBranchExist(r.refName) {
+		r.commit, err = r.repo.GetBranchCommit(r.refName)
+		if err != nil {
+			ctx.ServerError("GetBranchCommit", err)
+			return nil
+		}
+	} else if r.repo.IsTagExist(r.refName) {
+		r.commit, err = r.repo.GetTagCommit(r.refName)
+		if err != nil {
+			ctx.ServerError("GetTagCommit", err)
+			return nil
+		}
+	} else if len(r.refName) >= 4 && len(r.refName) <= 40 {
```

Comment:
We should filter values like `heads/pulls/345`. There's a Regexp somewhere that checks if a string is a valid commit sha1 (`^[0-9a-f]{4,40}$`).

Reply:
I went ahead and went with {4,64} to remove the need to change it if one manages to use gitea with a sha256-enabled repo, but will reduce that back to 40 on request...